### PR TITLE
Implement chain publisher

### LINKED ISSUE #1: notes
* factory: remove support for all other types except for aggregator
* consumer: remove support for filter array, only 1 filter is accepted
* split `lastHead` to each consumer
* maxRange logic is no longer necessary (it's original designed for chart factory)
* change database interface to a simple key-value db (localstorage is suitable for this)
* worker only need to process head (there's no past processor)

```
event = {
   filter,
   applyLogs: (storage, logs) => {
     ....
     return {
         key: newValue,
         key: newValue,
         ... all state changes ...
     },
   },
   safeDepth,
}
```

Storage and Logic: `processLogs(event, storage, logs, fromBlock, toBlock)`
* safeBlock = toBlock - safeDepth
* changes = event.applyLogs(storage, logs.filter(blockNumber <= safeBlock))
* store all changes to storage
* (lazy in-memory evaluation) unsafeChanges = event.applyLogs(storage, logs.filter(blockNumber > safeBlock))
* set lastHead to safeBlock

Notes:
* `unsafeChanges` temporary applied to current storage when client queries for states

### PR DESCRIPTION
**Pull request by CaoThien96**
To fix #1

### CODE REVIEW

File: test/quick-start/SfarmExample.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            address: BANK_ADDRESS,
+            topics: [
+                "0x1464b325679b3d6e47425f6f128fa127a13e437cd0614dc73559bf67ab24f08d",
+            ]
+        },
+        genesis,
+        applyLogs: async (storage, logs) => {
+            const state = await storage.getItem('sfarm-tokens') || {
+                range: genesis - 1,
+                changes: {}
```
Comment (Zergity) [inline test/quick-start/SfarmExample.js:95]: why do we need to store `changes` here?

File: test/quick-start/SfarmExample.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    },
+    {
+        key: 'sfarm-balances-in',
+        filter: filters[1],
+        genesis,
+        applyLogs: async (storage, logs) => {
+            const changes = await applyLogsBalance(storage, logs, 'sfarm-balances-in')
+            return changes
+        },
+        safeDepth: 64,
```
Comment (Zergity) [inline test/quick-start/SfarmExample.js:133]: try the sample with different `safeDepth` for each consumer

File: test/quick-start/SfarmExample.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        order: 1
```
Comment (Zergity) [inline test/quick-start/SfarmExample.js:134]: please remove the order logic, and call the subscriber.appLogs consequently

File: lib/chain-publisher.js [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        this._startProcessingLoop()
+    }
+
+    _getProvider = (seed) => {
+        const i = Math.floor(this.config?.provider.length * (seed ?? Math.random()))
+        return this.config?.provider[i]
+    }
+
+    async _startProcessingLoop() {
+        let nextPastProcess = Date.now()
```
Comment (Zergity) [inline lib/chain-publisher.js:28]: `nextPastProcess` is unused

**Review by Zergity — CHANGES_REQUESTED** [review]

File: test/quick-start/SfarmExample.js (hunk 1 above)
Comment (CaoThien96) [reply test/quick-start/SfarmExample.js:95]: I do not understand. `changes` is value that the client saves after processing the logs

File: lib/chain-publisher.js (hunk 4 above)
Comment (CaoThien96) [reply lib/chain-publisher.js:28]: `nextPastProcess` Used in next loop when `this.context.lastProcessed == this.context.head`, will wait 3 seconds. The intent is that every time the processor repeats with a different head

File: lib/chain-publisher.js [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                const requests = requestsByKey[key]
+
+                if (!requests || !requests.length) {
+                    return
+                }
+
+                const consumerLogs = _.chain(requests)
+                    .map(request => filterLogs(logs, request))
+                    .flatten()
+                    .sort((a, b) => (a.blockNumber - b.blockNumber) || (a.logIndex - b.logIndex))
```
[the remark this answers is not in the document]
Comment (CaoThien96) [reply lib/chain-publisher.js]: resolved

File: test/quick-start/SfarmExample.js (hunk 1 above)
Comment (CaoThien96) [reply test/quick-start/SfarmExample.js]: resolved

File: test/quick-start/SfarmExample.js (hunk 2 above)
Comment (CaoThien96) [reply test/quick-start/SfarmExample.js]: resolved

File: test/quick-start/SfarmExample.js (hunk 3 above)
Comment (CaoThien96) [reply test/quick-start/SfarmExample.js]: resolved

File: doc/quick-start.md [hunk 6]
```diff
@@ -6,34 +6,23 @@
...
 # Step 5.
 #
 # Test it by run for a while.
-timeout 15 node worker.js
+timeout 15s node WalletSample.js
 ```
 
-* File [sfarm-abi.json](../test/quick-start/sfarm-abi.json)
-* File [worker.js](../test/quick-start/worker.js)
+* File [WalletSample.js](../test/quick-start/WalletSample.js)
```
Comment (Zergity) [inline doc/quick-start.md:28]: why test files is located in a `quick-start` folder?

Comment (CaoThien96) [reply doc/quick-start.md:28]: I misunderstood the purpose of quick start. 
It should be SfarmExample.js which would be more correct

Comment (Zergity) [reply doc/quick-start.md:28]: I mean the SfarmExample.js, why is it INSIDE the folder `test/quick-start`?

**Review by Zergity — CHANGES_REQUESTED** [review]
please change the apllyLogs interface to: `function applyLogs(storage, logs)` and handle the `storage.setItem` and `removeItem` in the lib

**Review by Zergity — APPROVED** [review]

**Pull request merged**